### DISCUSSION TITLE
V_VALS_C and E_VALS_C

### DISCUSSION THREAD

**Discussion by gdalle** [simonschoelly/SimpleValueGraphs.jl#21] [Ideas]
Hi there,

Following up on our discussion in Slack :) First of all, thanks for your work on this library!

I understand the purpose of these two types is to describe the storage of vertex and edge values. However, since the related fields are [constructed](https://github.com/simonschoelly/SimpleValueGraphs.jl/blob/5cd233befdc2c95500050c5d45f88cabe23a40e6/src/valuegraph.jl#L179) using `typeof(...)`, I'm not sure the compiler is able to infer the types correctly. Here is an example supporting my hunch:

```julia
julia> using SimpleValueGraphs

julia> using Test

julia> @inferred ValGraph(0)
ERROR: return type ValGraph{Int32,Tuple{},Tuple{},Tuple{},Tuple{},Tuple{}} does not match inferred return type ValGraph{Int32,Tuple{},Tuple{},Tuple{},Tuple{},_A} where _A
Stacktrace:
 [1] error(::String) at ./error.jl:33
 [2] top-level scope at none:1
```

What do you think? Could we make this better somehow?

**Comment by simonschoelly** [comment]
So it is indeed true that the compiler can't infer the type statically. In this case, this can certainly be fixed, maybe with usage of the`@generated` macro. What might probably be a bit more difficult are constructors of the form
```julia
ValGraph(0, edgeval_types=..., vertexval_types)
```
as it it is not always clear when keyword arguments are resolved at at runtime. I think this relies also a bit on const propagation, where there are no clear rules when that optimization is used. If it turns out that it is not possible to make the compiler infer the type statistically,  there can always be the alternative constructor of the form
```julia
ValGraph{V, V_VALS, E_VALS, G_VALS}
```
This notation might just be a bit awkward to write.

In any case, I don't think it is a big issue here, as when you call some function with your graph, i.e. `f(g, ...)`, the type of the graph is fully determined inside this function.

There is another issue with type instability, and that is when you have a graph with multiple vertex values (or edge, graph values) of different types, then `get_vertexval(g, v, 1)` is not typestable anymore. This can maybe be resolved in the future by inlining functions, in the hope that const propagation is triggered.

**Reply by gdalle** [reply]
I'm not sure what you mean by

> In any case, I don't think it is a big issue here, as when you call some function with your graph, i.e. f(g, ...), the type of the graph is fully determined inside this function.

I'm just a Julia newbie but if the compiler cannot infer the full parametric graph type statically, how can it compile the best possible version of the function applied to `g`?
Do you have a source on constant propagation? It doesn't feature in the official Julia manual...

While we're at it, is there a specific reason why you chose to store the vertex and edge values as Tuples of Vectors rather than Vectors of Tuples?

**Reply by gdalle** [reply]
Actually, why not have arbitrary structs as vertex and edge vals, is there a performance benefit to NamedTuples? (genuinely curious here)

**Reply by gdalle** [reply]
For instance, when I needed to implement a valued graph for myself, I did it like this, and now you make me wonder what's wrong with it :) any advice will be appreciated

```julia
mutable struct OrderedLabeledDiGraph{VT, ET} <: AbstractGraph{Int}
    nv::Int
    edges::Vector{DumbEdge}
    inneighbors::Vector{Vector{Int}}
    outneighbors::Vector{Vector{Int}}
    edgeindex::Vector{Vector{Int}}
    vertexlabels::Vector{VT}
    edgelabels::Vector{ET}
end
```

**Reply by simonschoelly** [reply]
Julias compiler works at runtime (bit similar to Javas JIT - but its not the same) - if it encounters some function that is called with some types it has not seen before, it compiles this code at runtime. That is why, when benchmarking Julia functions, the first run often is much slower.

Only source I have on constant propagation are some threads from discourse: https://discourse.julialang.org/search?q=constant%20propagation

There is nothing wrong with your graph type, although you might consider generating the edges with an iterator instead of storing them all the time.

In your case,  you might think about how you would proceed when inserting a new edge. Is `edgelabels` sorted?  Then you might have to do a big shift in that array. If it is not sorted, you can just append it to the end of the `edgelabels`  but when you iterate over outgoing edges of some vertex (which happens in a lot of graph algorithms) and also want to retrieve the edge values for these edges, you will have a lot of cache misses/page faults.

The reason why I choose tuple-of-vectors instead of vector-of-tuples is that when an algorithm only needs access to to one edge value, all data lies more packed in cache. The downside is that the code is more complex, and adding edges or changing values might also be slightly less performant. See also https://en.wikipedia.org/wiki/AoS_and_SoA.

This holds only for the concrete graph types implement in this package, `AbstractValGraph` does not say that one can not have a vector of tuples. Here is an example where, at least for the moment, I use vector-of-tuples: https://github.com/simonschoelly/GraphDatasets.jl/blob/master/src/valgraphcollection.jl

It would have been possible to not only allow tuples and named tuples, but structs in general. But I thought that dealing with tuples is already complicated enough and if someone wants to store structs, they can always generate a single valued tuple that contains this struct. <https://github.com/JuliaArrays/StructArrays.jl> is an example of a package, that allows one to store arbitrary structs as struct-of-arrays.

**Reply by gdalle** [reply]
OK, I think I understand now why my initial "problem" doesn't really matter :) thanks for your detailed explanations!

**Reply by gdalle** [reply]
I didn't know about inlining but I looked it up, and it seems like a nice solution to the `get_value` type instability. What is stopping you from implementing it? A lack of benchmarks?

**Reply by simonschoelly** [reply]
Just some lack of time - I only found really time in the last few months to get this package into the shape it is now. Actually I also now merged some patch that will help with type instability there.

**Reply by gdalle** [reply]
Hey! I'm getting back into graph structures, and I was wondering if you had made some progress towards solving type instability within `get_vertexval(g, v, 1)`? If not, can we guess how much it affects performance?

**Reply by simonschoelly** [reply]
I have not really had time to work on this yet, I probably would also have to check if constant propagation works not better on Julia v1.7.0.

It is rather heard to say how much performance is impacted, it might depend on how many different types are used for `V_VALS` and the code surrounding that statement.

**Comment by simonschoelly** [comment]
Btw. I now fixed some of the constructor instability but its only on the master branch.
